### PR TITLE
Fix #812: New component: CodeScanner

### LINKED ISSUE #140: New component: code scanner
Based on https://github.com/serratus/quaggaJS

### PR DESCRIPTION
**Pull request by jepsar**
Fix https://github.com/primefaces-extensions/primefaces-extensions.github.com/issues/812

### CODE REVIEW

File: src/main/resources/META-INF/resources/primefaces-extensions/codescanner/codescanner-widget.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                    return new ZXing.BrowserBarcodeReader();
+                case 'qr':
+                    return new ZXing.BrowserQRCodeReader();
+                default:
+                    return new ZXing.BrowserMultiFormatReader();
+            }
+        }
+
+        function handleResult(result) {
+            if ($this.hasBehavior('codeScanned')) {
```
Comment (jepsar) [inline src/main/resources/META-INF/resources/primefaces-extensions/codescanner/codescanner-widget.js:27]: @melloware I'm currently in the process of testing / adding details, but I fail to get this part working. Do you know how I should get the 'codeScanned' behavior to be available?

File: src/main/java/org/primefaces/extensions/component/codescanner/CodeScannerRenderer.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            writer.writeAttribute("height", codeScanner.getHeight(), null);
+        }
+        writer.endElement("video");
+    }
+
+    protected void encodeScript(final FacesContext context, final CodeScanner codeScanner) throws IOException {
+        final WidgetBuilder wb = getWidgetBuilder(context);
+        wb.init("ExtCodeScanner", codeScanner.resolveWidgetVar(), codeScanner.getClientId(context));
+        wb.attr("type", codeScanner.getTypeEnum().name());
+        wb.finish();
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/component/codescanner/CodeScannerRenderer.java:84]: You need `encodeClientBehaviors` here in the widget creation before you call 'finish`.

File: src/main/java/org/primefaces/extensions/component/codescanner/CodeScannerRenderer.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Renderer for the {@link CodeScanner} component.
+ *
+ * @author Jasper de Vries &lt;jepsar@gmail.com&gt;
+ * @since 8.0.5
+ */
+public class CodeScannerRenderer extends CoreRenderer {
+
+    @Override
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/component/codescanner/CodeScannerRenderer.java:34]: You need a `decode` method with at least `decodeClientBehaviors`

**Review by melloware — CHANGES_REQUESTED** [review]

File: src/main/java/org/primefaces/extensions/component/codescanner/CodeScannerRenderer.java (hunk 3 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/component/codescanner/CodeScannerRenderer.java:34]: Ahhhh.. I trashed too much. I removed a bunch of code as I don't have a value to decode. Thanks!

Comment (melloware) [reply src/main/java/org/primefaces/extensions/component/codescanner/CodeScannerRenderer.java:34]: dont forget the encodeClientBehaviors too in the encodeScript method.

**Comment by jepsar** [comment]
Okay, this works as a base. I'll continue to work on stuff like:

* Start / stop
* Stop after successful scan
* Video boolean (scan without video element)
* For attribute, pointing to an input to copy the scanned code to
* DeviceId attribute

... and of course the showcase.

But I'll add them as separate PR's don't you agree?

File: src/main/java/org/primefaces/extensions/component/codescanner/CodeScanner.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    public enum PropertyKeys {
+        widgetVar,
+        type,
+        style,
+        styleClass,
+        width,
+        height
+    }
+
+    public enum Type {
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/component/codescanner/CodeScanner.java]: Type seems too generic a name for an enumeration doesn't it?

**Review by melloware — CHANGES_REQUESTED** [review]

File: src/main/java/org/primefaces/extensions/component/codescanner/CodeScanner.java (hunk 4 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/component/codescanner/CodeScanner.java]: Agreed

**Pull request merged**